src/mlpack/tests/scd_test.cpp

```diff
@@ -0,0 +1,165 @@
+#include "test_tools.hpp"
+
+using namespace std;
+using namespace mlpack;
+using namespace mlpack::math;
+using namespace mlpack::optimization;
+using namespace mlpack::optimization::test;
+using namespace mlpack::regression;
+
+BOOST_AUTO_TEST_SUITE(SCDTest);
+
+/**
+ * Test the correctness of the SCD implementation by using a dataset with a
+ * precalculated minima.
+ */
+BOOST_AUTO_TEST_CASE(PreCalcSCDTest)
+{
+  arma::mat predictors("0 0 0.4; 0 0 0.6; 0 0.3 0; 0.2 0 0; 0.2 -0.5 0;");
+  arma::Row<size_t> responses("1  1  0;");
+
+  LogisticRegressionFunction<arma::mat> f(predictors, responses, 0.0001);
+
+  SCD<> s(0.02, 60000, 1e-5);
+  arma::mat iterate = f.InitialPoint();
+
+  double objective = s.Optimize(f, iterate);
+
+  BOOST_REQUIRE_LE(objective, 0.055);
+}
+
+/**
+ * Test the correctness of the SCD implemenation by using the sparse test
+ * function, with dijoint features which optimize to a precalculated minima.
+ */
+BOOST_AUTO_TEST_CASE(DisjointFeatureTest)
+{
+  // The test function for parallel SGD should work with SCD, as the gradients
+  // of the individual functions are projections into the ith dimension.
+  SparseTestFunction f;
+  SCD<> s(0.4);
+
+  arma::mat iterate = f.GetInitialPoint();
+
+  double result = s.Optimize(f, iterate);
+
+  // The final value of the objective function should be close to the optimal
+  // value, that is the sum of values at the vertices of the parabolas.
+  BOOST_REQUIRE_CLOSE(result, 123.75, 0.01);
+
+  // The co-ordinates should be the vertices of the parabolas.
+  BOOST_REQUIRE_CLOSE(iterate[0], 2, 0.02);
+  BOOST_REQUIRE_CLOSE(iterate[1], 1, 0.02);
+  BOOST_REQUIRE_CLOSE(iterate[2], 1.5, 0.02);
+  BOOST_REQUIRE_CLOSE(iterate[3], 4, 0.02);
+}
+
+/**
+ * Test the greedy descent policy.
+ */
+BOOST_AUTO_TEST_CASE(GreedyDescentTest)
+{
+  // In the sparse test function, the given point has the maximum gradient at
+  // the feature with index 2.
+  arma::mat point("1; 2; 3; 4;");
+
+  SparseTestFunction f;
+
+  GreedyDescent descentPolicy;
+
+  BOOST_REQUIRE_EQUAL(descentPolicy.DescentFeature(0, point, f), 2);
+
+  point[1] = 10;
+
+  BOOST_REQUIRE_EQUAL(descentPolicy.DescentFeature(0, point, f), 1);
+}
+
+/**
+ * Test changes to Logistic regression function.
+ */
+BOOST_AUTO_TEST_CASE(LogisticRegressionFunctionFeatureGradientTest)
+{
+  // Evaluate the gradient and feature gradient and equate.
+  arma::mat predictors("0 0 0.4; 0 0 0.6; 0 0.3 0; 0.2 0 0; 0.2 -0.5 0;");
+  arma::Row<size_t> responses("1  1  0;");
+
+  LogisticRegressionFunction<arma::mat> f(predictors, responses, 0.0001);
+
+  arma::mat testPoint(1, f.NumFeatures(), arma::fill::randu);
+
+  arma::mat testGradient;
+
+  f.Gradient(testPoint, testGradient);
+
+  for (size_t i = 0; i < f.NumFeatures(); ++i)
+  {
+    arma::sp_mat fGrad;
+    f.FeatureGradient(testPoint, i, fGrad);
+
+    CheckMatrices(testGradient.col(i), arma::mat(fGrad.col(i)));
+  }
+}
+
+/**
+ * Test changes to Softmax regression function.
```

Comment:
I think this might be an inaccurate comment (or at least it won't make sense after merge), maybe something more like `Test that SoftmaxRegressionFunction::FeatureGradient()` works as expected.